### DISCUSSION TITLE
Assign function from one space into mixed space

### DISCUSSION THREAD

**Discussion by Hasupama** [firedrakeproject/firedrake#3322] [Firedrake support] [answered]
Hi, 
I want to assign the values of a function defined in one space to mixed function space as the example below. The assign function did not work. Which function should I need to use in this case to assign the 'phi1' values to 'phi'? 

```
solver_parameters ={	"pc_factor_mat_solver_type": "mumps",
                          "snes_rtol": 1.e-4,
                          "snes_atol": 1.e-3,  
                     	 "maximum_iterations": 100,
                    		"report": True,
                     	 	"error_on_nonconvergence": True,
                     	 									 }


mesh = Mesh('domain.msh')
x = SpatialCoordinate(mesh)

Space_phi = FunctionSpace(mesh, 'CG', 1)
dphi = TrialFunction(Space_phi)
del_phi = TestFunction(Space_phi)
phi1 = Function(Space_phi)

F_phi = x[0]*(  8.854187E-12*phi1.dx(i)*del_phi.dx(i)  )*dx
bc_list=[]
bc_list.append( DirichletBC(Space_phi.sub(0), 0.0, 12) )
bc_list.append( DirichletBC(Space_phi.sub(0), 20e3, 11) )

problem = NonlinearVariationalProblem(F_phi, phi1, bcs=bc_list,   form_compiler_parameters={"quadrature_degree":2})
solver = NonlinearVariationalSolver(problem)
solver.parameters.update(solver_parameters)
solver.solve()

scalar_lin_space = FunctionSpace(mesh, 'CG', 1)
Space = MixedFunctionSpace([scalar_lin_space,scalar_lin_space])
u = Function(Space)
phi,c = split(u)
phi.assign(phi1)
```

**Comment by dham** [comment]
This is the easy to make confusion about what `split` does. `split` symbolically indexes the entries of the function, which is what you want if you are going to use the components in a Form. However, to actually do something numerical such as assign you instead need the subfunctions of `u`:
```python
phi, c = u.subfunctions
```

**Reply by Hasupama** [reply]
When I tried with subfunction instead of split it gives the error
AttributeError: 'CoordinatelessFunction' object has no attribute 'subfunctions'

The phi values contribute for the weak form. 

Is it ok to use assign function for this case I need to assign values which are in different FunctionSpaces.

**Reply by dham** [reply, marked as answer]
Here is a minimal example:
```python
mesh = UnitSquareMesh(10, 10)
x = SpatialCoordinate(mesh)

Space_phi = FunctionSpace(mesh, 'CG', 1)
phi1 = Function(Space_phi)

scalar_lin_space = FunctionSpace(mesh, 'CG', 1)
Space = MixedFunctionSpace([scalar_lin_space,scalar_lin_space])
u = Function(Space)
phi,c = u.subfunctions
phi.assign(phi1)
```
I have verified that this works.

If you then need to use `phi` in a weak form then you should redefine it using `split` before you do so.

Assign only works between matching function spaces. If you need to transfer Function data between different function spaces then you either need to interpolate or project.
